### DISCUSSION TITLE
New YouTube video showing undetected-chromedriver with bug-fixes and enhancements

### DISCUSSION THREAD

**Discussion by mdmintz** [ultrafunkamsterdam/undetected-chromedriver#1636] [Show your project! using undetected-chromedriver]
New YouTube video showing undetected-chromedriver with bug-fixes and enhancements:
<b>https://www.youtube.com/watch?v=5dMFI3e85ig</b>

<a href="https://www.youtube.com/watch?v=5dMFI3e85ig"><img src="https://github.com/seleniumbase/SeleniumBase/assets/6788579/4c9a112a-25c6-458f-83c6-c4b910b8b636" title="undetected automation" width="330" /></a>

Note the improvements made:
* Driver version-detection & management.
* Allowing mismatched browser/driver versions.
* Automatically changing the user agent to prevent detection. (`HeadlessChrome` to `Chrome`)
* Automatically disconnecting chromedriver from Chrome as needed.
* Multithreading tests in parallel via `pytest-xdist`.
* Adjusting configuration based on the environment.
* Has options for setting proxy and proxy-with-auth.

Many of these fixes can be back-ported easily to undetected-chromedriver. (Explanation in the video of how to do that, such as using `requests.get(url)` before navigating to a URL to know if you need to disconnect/reconnect the driver from Chrome.)

Basic script:

```python
from seleniumbase import Driver
import time

driver = Driver(uc=True)
driver.get("https://nowsecure.nl/#relax")
time.sleep(4)
driver.quit()
```

Here's a more-advanced script with retries and a captcha-click failsafe for bypassing detection:

```python
from seleniumbase import SB

with SB(uc=True) as sb:
    sb.driver.get("https://nowsecure.nl/#relax")
    sb.sleep(1)
    if not sb.is_text_visible("OH YEAH, you passed!", "h1"):
        sb.get_new_driver(undetectable=True)
        sb.driver.get("https://nowsecure.nl/#relax")
        sb.sleep(1)
    if not sb.is_text_visible("OH YEAH, you passed!", "h1"):
        if sb.is_element_visible('iframe[src*="challenge"]'):
            with sb.frame_switch('iframe[src*="challenge"]'):
                sb.click("span.mark")
                sb.sleep(2)
    sb.activate_demo_mode()
    sb.assert_text("OH YEAH, you passed!", "h1", timeout=3)
```

**Comment by joe-box** [comment]
hello mdmintz
I do some code with seleniumbase, it is great
i checked that uc can not bypass cloudflare now but seleniumbase(uc mode) can do it 
and i want to know how to use seleniumbase debugging and how to remote a opened chrome browser with '--remote-debugging-port='
thank you so much

**Reply by mdmintz** [reply]
Hello,
For remote debugging: If you use `--remote-debug` / `remote_debug=True`, then you should be able to see those pages from `chrome://inspect/#devices` and `chrome://inspect/#pages`

<img width="620" alt="Screenshot 2023-10-29 at 9 52 32 AM" src="https://github.com/ultrafunkamsterdam/undetected-chromedriver/assets/6788579/ecf3b62f-3bcf-4875-9e89-91b00c5d1dd8">

For command-line debugging, use a `breakpoint()` in the location where you want a breakpoint. That takes you into the Pdb+ debugger. (`n`: next, `c`: continue, `s`: step into)

**Comment by nozonyan** [comment]
I'm trying to adapt my current UC script to sbase w/ uc mode on, but it seems I cannot override download folders using sbase. Am I doing it wrong or sbase does not allow this? Because it is sadly essential to me that the files are downloaded to separated folders for proper bibliographic organization.

**Reply by mdmintz** [reply]
The methods exist:

```python
self.download_file(file_url, destination_folder=None)

self.save_file_as(file_url, new_file_name, destination_folder=None)
```

**Comment by jdholtz** [comment]
This framework works great. I highly recommend everyone using undetected-chromedriver to switch over to SeleniumBase, as it is still maintained and works very well.
